README.md
### What is this repository for? ###

This is the home of SeekMIDI, a simple graphical MIDI sequencer offering multi-channel MIDI sequencing. Look at [Seq24](https://launchpad.net/seq24/) for a slightly heavier example of what we mean.
Written in Perl, SeekMIDI uses GTK+3 to provide its GUI and MIDI-Perl to provide its MIDI handling system. Please download and try it.
For the latest features, you can also use git to download the latest master. We aim to keep master somewhat stable so as not to have to release as often.

__Please keep in mind that SeekMIDI is still alpha software and features are constantly being implemented.__

### How do I get set up? ###

See [the wiki](https://github.com/oldtechaa/SeekMIDI/wiki/HowTo:-Install-and-Run).

We will soon have a full automated setup procedure for SeekMIDI. There are no plans to port it to Windows or OS X at this moment, although a third-party port would not be discouraged.

### Contribution guidelines ###

To contribute, please take a look at the ["help-wanted" issues](https://github.com/oldtechaa/SeekMIDI/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22). Alternatively, you can email me, and we can find a task suitable to your wishes and experience. We welcome contributions and newcomers!

### Who do I talk to? ###

It's just me. [oldtechaa@gmail.com](mailto:oldtechaa@gmail.com)

### The main website ###

The main website's address currently is [the Github repo](https://github.com/oldtechaa/SeekMIDI/).

### Where's the documentation? ###

The documentation can be found [here at our wiki](https://github.com/oldtechaa/SeekMIDI/wiki).


CHANGELOG.md
# Change Log
Notable changes will be documented here. This changelog follows the format from [Keep a Changelog](http://keepachangelog.com/).

## [Unreleased]
### Added
- Patch change support.
- Depends on Readonly for constants.

### Changed
- Switch to using GTK+3 instead of GTK+2.
- Save command has been moved into menus instead of a button in the main interface.
- Save command uses native GTK+ save dialog instead of file entry.

### Fixed
- Bug fix for issue #30, main widget now draws fully to the edges of the widget area.

## [0.2.1-alpha] - 2016-10-08
### Added
- Measure markings are provided at the top and throughout the note grid.
- README now links to the wiki.
- Allow changing volume of notes.
- Extend CONTRIBUTING.md to be more meaningful.

### Fixed
- Bug fix for issue #18, now you can run SeekMIDI outside of the directory containing the script.
- Bug fix for issue #11, which spews out a bunch of unnecessary warnings.
- Fix for issue #15: higher notes are now at the top of the plotting area.
- Fix #13: An empty filename now does nothing.

### Changed
- Left-click and drag now adds notes, right-click removes.
- No longer rely on installation of MIDI-Perl. It is now included in the source directory under lib/.
- SeekMIDI.pl can now be found in the bin/ directory.

## 0.1.0-alpha - 2016-03-25
### Added
- First release. Added the last bit of code for getting the MIDI output from the GUI.

[Unreleased]: https://github.com/oldtechaa/SeekMIDI/compare/v0.2.1-alpha...HEAD
[0.2.1-alpha]: https://github.com/oldtechaa/SeekMIDI/compare/v0.1.0-alpha...v0.2.1-alpha


MIDway.pl
#!/usr/bin/perl

# I don't know if this is a legally binding copyright, but
# Copyright 2016 oldtechaa (Easily found on the Internet.)
# This software is licensed under the Perl license. (That means Artistic v1 or GPLv1.)

use strict;
use warnings;

use MIDI;

# writes the MIDI output to a file based on the list of events and the filename
sub midiWrite {
	my $midiEventsRef = shift;
	my $midiTicks = shift;
	my $midiFile = shift;
	
	my $midiTrack = MIDI::Track->new({'events' => $midiEventsRef});
	my $midiPiece = MIDI::Opus->new({'format' => 0, 'ticks' => $midiTicks, 'tracks' => [$midiTrack]});
	$midiPiece->write_to_file($midiFile);
};

my $events = [
	['patch_change',    0, 0,  19],
	['patch_change',    0, 1, 124],
	[     'note_on',    0, 0,  50, 64],
	[     'note_on',    0, 1,  50, 96],
	[    'note_off',   96, 0,  50, 64],
	[     'note_on',   96, 0,  49, 64],
	[    'note_off',   96, 1,  50, 96],
	[     'note_on',   96, 1,  49, 90],
	[    'note_off',  192, 0,  48, 64],
	[     'note_on',  192, 0,  49, 64],
	[    'note_off',  192, 1,  48, 90],
	[     'note_on',  192, 1,  49, 90],
	[    'note_off',  288, 0,  48, 64],
	[     'note_on',  288, 0,  49, 64],
	[    'note_off',  288, 1,  48, 90],
	[     'note_on',  288, 1,  49, 90],
	[    'note_off',  384, 0,  48, 64],
	[     'note_on',  384, 0,  49, 64],
	[    'note_off',  384, 1,  48, 90],
	[     'note_on',  384, 1,  49, 90],
	[    'note_off',  480, 0,  48, 64],
	[     'note_on',  480, 0,  49, 64],
	[    'note_off',  480, 1,  48, 90],
	[     'note_on',  480, 1,  49, 90],
	[    'note_off',  576, 0,  48, 64],
	[     'note_on',  576, 0,  49, 64],
	[    'note_off',  576, 1,  48, 90],
	[     'note_on',  576, 1,  49, 90],
	[    'note_off',  672, 0,  48, 64],
	[     'note_on',  672, 0,  49, 64],
	[    'note_off',  672, 1,  48, 90],
	[     'note_on',  672, 1,  49, 90],
	[    'note_off',  768, 0,  48, 64],
	[     'note_on',  768, 0,  49, 64],
	[    'note_off',  768, 1,  48, 90],
	[     'note_on',  768, 1,  49, 90],
	[    'note_off',  864, 0,  48, 64],
	[     'note_on',  864, 0,  49, 64],
	[    'note_off',  864, 1,  48, 90],
	[     'note_on',  864, 1,  49, 90],
	[    'note_off',  960, 0,  48, 64],
	[     'note_on',  960, 0,  49, 64],
	[    'note_off',  960, 1,  48, 90],
	[     'note_on',  960, 1,  49, 90],
	[    'note_off', 1056, 0,  48, 64],
	[    'note_off', 1056, 1,  48, 90]
];

midiWrite($events, 1056,"Milestone1Test.mid");


SeekMIDI.pl
#!/usr/bin/perl

# SeekMIDI, a simple graphical MIDI sequencer
# Copyright (C) 2016  oldtechaa  <oldtechaa@gmail.com>
# Version 0.1.0

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;

# custom widget class; separate from main package below
package Gtk2::MIDIPlot;

use Gtk2;
use base 'Gtk2::VBox';
use Cairo;
use Pango;

# makes a class-global array that holds true/false values for which note blocks are enabled, and the global drawing area
my @gtkObjects;

my $this;
my $thisScroll;
my $volSlider;
my $HBox;

my ($dragRow, $dragStart, $dragMode) = (-1, -1, -1);

my @notes = (0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
             0, 1, 0, 1, 0, 0, 1, 0);

# sets up the class; asks for the signals we need; sets main widget size
sub new {
  $this = Gtk2::DrawingArea->new();
  my $topVBox = bless Gtk2::VBox->new();
  $thisScroll = Gtk2::ScrolledWindow->new();
  $HBox = Gtk2::HBox->new();
  my $volLabel = Gtk2::Label->new('Volume: ');
  $volSlider = Gtk2::HScale->new_with_range(0, 127, 1);

  $topVBox->pack_start($thisScroll, 1, 1, 0);
  $topVBox->pack_start($HBox, 0, 0, 0);

  $thisScroll->add_with_viewport($this);

  $HBox->pack_start($volLabel, 0, 0, 0);
  $HBox->pack_start($volSlider, 1, 1, 0);

  $this->signal_connect(expose_event => 'Gtk2::MIDIPlot::expose');
  $this->signal_connect(button_press_event => 'Gtk2::MIDIPlot::button');
  $this->signal_connect(button_release_event => 'Gtk2::MIDIPlot::release');
  $this->signal_connect(motion_notify_event => 'Gtk2::MIDIPlot::motion');
  $thisScroll->get_hadjustment->signal_connect(value_changed => 'Gtk2::MIDIPlot::expose');
  $thisScroll->get_vadjustment->signal_connect(value_changed => 'Gtk2::MIDIPlot::expose');

  # ask for mouse events from the DrawingArea
  $this->set_events(['button-press-mask', 'button-motion-mask', 'button-release-mask']);

  $this->set_size_request(28836, 1040);

  # handles initializing the @gtkObjects array
  for(my $incx = 0; $incx < 2400; $incx++) {
    for(my $incy = 0; $incy < 128; $incy++) {
      $gtkObjects[$incx][$incy] = 0;
    }
  }

  return($topVBox);
}

# refresh handler; handles drawing grid and objects
sub expose {
  $this->window->clear();

  # makes new Cairo context
  my $thisCairo = Gtk2::Gdk::Cairo::Context->create($this->get_window());

  # sets drawing color for main grid
  $thisCairo->set_source_rgb(0.75, 0.75, 0.75);

  # get the current scroll positions and size of the window, then convert to grid-blocks, adjusting to draw surrounding blocks also, and make sure we don't go out of bounds
  my ($xmin, $ymin, $width, $height) = (int($thisScroll->get_hadjustment()->value / 12) + 3, int($thisScroll->get_vadjustment()->value / 8) + 2, $thisScroll->get_hadjustment()->page_size - 36, $thisScroll->get_vadjustment()->page_size - 16);
  my $xmax = ($xmin + (int($width / 12) + 2));
  my $ymax = ($ymin + (int($height / 8) + 2));
  if($xmax > 2403) {$xmax = 2403};
  if($ymax > 130) {$ymax = 130};

  # these two loops create the background grid
  my $inc;
  for ($inc = $xmin; $inc <= $xmax; $inc++) {
    $thisCairo->move_to($inc * 12, $ymin * 8);
    $thisCairo->line_to($inc * 12, $ymax * 8);
  }
  for ($inc = $ymin; $inc <= $ymax; $inc++) {
    $thisCairo->move_to($xmin * 12, $inc * 8);
    $thisCairo->line_to($xmax * 12, $inc * 8);
  }

  # the grid must be drawn before we start redrawing the note objects
  $thisCairo->stroke();

  $thisCairo->set_source_rgb(0, 0, 0);
  for($inc = $ymin; $inc < $ymax; $inc++) {
    if($notes[127 - ($inc - 2)] == 1) {
      $thisCairo->rectangle(($xmin - 3) * 12, $inc * 8 + 1, 32, 6);
    }
  }

  my $thisPango = Pango::Cairo::create_layout($thisCairo);
  my $thisPangoAttr = Pango::AttrList->new();
  $thisPangoAttr->insert(Pango::AttrSize->new(8192));
  $thisPango->set_attributes($thisPangoAttr);
  for($inc = $xmin - 3; $inc < $xmax - 2; $inc++) {
    if($inc % 4 == 0) {
      $thisPango->set_text($inc * 12);
      my ($PangoWidth, $PangoHeight) = $thisPango->get_size();
      $thisCairo->move_to($inc * 12 - $PangoWidth / Gtk2::Pango->scale() / 2 + 36, $ymin * 8 - 12);
      Pango::Cairo::show_layout($thisCairo, $thisPango);
    }
  }

  # this checks for events with their state set to true, then draws them 
  for(my $incx = $xmin; $incx < $xmax; $incx++) {
    for(my $incy = $ymin; $incy < $ymax; $incy++) {
      if($gtkObjects[$incx - 3][$incy - 2] == 1) {
        $thisCairo->rectangle($incx * 12, $incy * 8, 12, 8);
      }
    }
  }

  # fill applies the black source onto the destination through the mask with rectangular holes
  $thisCairo->fill();
}

# handles mouse-clicks on the custom widget
sub button {
  my $event = $_[1];

  my ($xcell, $ycell) = (($event->x - ($event->x % 12)) / 12, ($event->y - ($event->y % 8)) / 8);
  my ($xmin, $ymin) = (int($thisScroll->get_hadjustment()->value / 12) + 3, int($thisScroll->get_vadjustment()->value / 8) + 2);

  # if the left mouse button then invert this gridbox's state value
  if ($event->button == 1) {
    if($xcell >= $xmin && $ycell >= $ymin) {
      $gtkObjects[$xcell - 3][$ycell - 2] = 1;

      # makes new Cairo context
      my $thisCairo = Gtk2::Gdk::Cairo::Context->create($this->get_window());

      $thisCairo->rectangle($xcell * 12, $ycell * 8, 12, 8);
      $thisCairo->fill();
      $dragMode = 1;

      # initialize drag variables
      if($dragStart == -1) {
        $dragStart = $xcell;
      }
      if($dragRow == -1) {
        $dragRow = $ycell;
      }
    }
  } elsif ($event->button == 3) {
    if($xcell >= $xmin && $ycell >= $ymin) {
      if($gtkObjects[$xcell - 3][$ycell - 2] == 1) {
        $gtkObjects[$xcell - 3][$ycell - 2] = 0;
        $dragMode = 0;
        expose();

        # initialize drag variables
        if($dragStart == -1) {
          $dragStart = $xcell;
        }
        if($dragRow == -1) {
          $dragRow = $ycell;
        }
      }
    }
  }
}

# handles mouse drag across the widget
sub motion {
  my $event = $_[1];

  my $xcell = ($event->x - ($event->x % 12)) / 12;

  # check if the underlying cell is set or not and if not, check which mouse button is pressed, then draw and set $gtkObjects
  if($dragMode == 1) {
    # makes new Cairo context
    my $thisCairo = Gtk2::Gdk::Cairo::Context->create($this->get_window());

    if($dragMode == 1) {
      # checks whether our overall drag is to the left or right and draws rectangles and updates $gtkObjects accordingly
      if($xcell >= $dragStart) {
        $thisCairo->rectangle($dragStart * 12, $dragRow * 8, ($xcell - $dragStart + 1) * 12, 8);
        for(my $inc = $dragStart; $inc <= $xcell; $inc++) {
          $gtkObjects[$inc - 3][$dragRow - 2] = 1;
        }
      } else {
        $thisCairo->rectangle($xcell * 12, $dragRow * 8, ($dragStart - $xcell) * 12, 8);
        for(my $inc = $xcell; $inc <= $dragStart; $inc++) {
          $gtkObjects[$inc - 3][$dragRow - 2] = 1;
        }
      }
    }
    $thisCairo->fill();
  } elsif($dragMode == 0) {
    # checks whether our overall drag is to the left or right and updates $gtkObjects accordingly
    if($xcell >= $dragStart) {
      for(my $inc = $dragStart; $inc <= $xcell; $inc++) {
        $gtkObjects[$inc - 3][$dragRow - 2] = 0;
      }
      expose($this);
    } else {
      for(my $inc = $xcell; $inc <= $dragStart; $inc++) {
        $gtkObjects[$inc - 3][$dragRow - 2] = 0;
      }
      expose($this);
    }
  }
}

# clears the current drag row, start point, and drag mode when the drag is ended
sub release {
  ($dragRow, $dragStart, $dragMode) = (-1, -1, -1);
}

sub getMIDI {
  my @events;

  push(@events, ['patch_change', 0, 0, 0]);

  my $delta = 0;
  for(my $incy = 0; $incy <= 127; $incy++) {
    if($gtkObjects[0][$incy] == 1) {
      push(@events, ['note_on', $delta, 0, 127 - $incy, 96]);
    }
  }
  $delta = 24;
  for(my $incx = 1; $incx < 2400; $incx++) {
    for(my $incy = 0; $incy < 127; $incy++) {
      if($gtkObjects[$incx][$incy] == 0 && $gtkObjects[$incx - 1][$incy] == 1) {
        push(@events, ['note_off', $delta, 0, 127 - $incy, 127]);
        $delta = 0;
      } elsif($gtkObjects[$incx][$incy] == 1 && $gtkObjects[$incx - 1][$incy] == 0) {
        push(@events, ['note_on', $delta, 0, 127 - $incy, 127]);
        $delta = 0;
      }
    }
    $delta = $delta + 24;
  }

  return \@events;
}

sub get_HBox {
  return $HBox;
}

package main;

# append to @INC the script dir's subdir 'lib' for MIDI.pm
use FindBin;
use File::Spec;
use lib File::Spec->catdir($FindBin::Bin, 'lib');

use MIDI;
use Gtk2 -init;
# use Locale::gettext;

# writes the MIDI output to a file based on the list of events and the filename
sub midiWrite {
	my $midiEventsRef = shift;
	my $midiTicks = shift;
	my $midiFile = shift;
	
	my $midiTrack = MIDI::Track->new({'events' => $midiEventsRef});
	my $midiPiece = MIDI::Opus->new({'format' => 0, 'ticks' => $midiTicks, 'tracks' => [$midiTrack]});
	$midiPiece->write_to_file($midiFile);
}

# reads from the event file (future capability, not sure if it will be added) This is just for testing, to reach early milestones without an event entry control.
# would mainly be useful if there was ever a CLI version or if the GUI version had the capability to read non-MIDI projects.
#sub evtOpen {
#  my $evtFile = shift;
#  my @events;
#
#  open(my $evtHandle, '<', $evtFile);
#
#  while (<$evtHandle>) {
#    push(@events, [split(/,\s*/, $_)]);
#  };
#
#  close($evtHandle);
#
#  return \@events;
#}

# creates window with title
my $window = Gtk2::Window->new();
$window->set_title('SeekMIDI MIDI Sequencer');

# creates VBox for widgets along the top and the main widget area below
my $mainVBox = Gtk2::VBox->new(0, 6);
$window->add($mainVBox);

# creates HBox for widgets along the top
my $controlHBox = Gtk2::HBox->new(0, 6);
$mainVBox->pack_start($controlHBox, 0, 0, 0);

# creates label for filename entry
my $fileLabel = Gtk2::Label->new('Output Filename:');
$controlHBox->pack_start($fileLabel, 0, 0, 0);

# creates filename entry field
my $fileEntry = Gtk2::Entry->new();
$controlHBox->pack_start($fileEntry, 0, 0, 0);

# creates file save button
my $saveButton = Gtk2::Button->new('_Save');
$controlHBox->pack_start($saveButton, 0, 0, 0);

# creates main widget
my $mainWidget = Gtk2::MIDIPlot->new();
$mainVBox->pack_start($mainWidget, 1, 1, 0);
$saveButton->signal_connect(clicked => sub{midiWrite($mainWidget->getMIDI(), 96, $fileEntry->get_text())});

# starts up the GUI
$window->signal_connect(destroy => sub{Gtk2->main_quit()});
$window->show_all();
$mainWidget->get_HBox()->hide();
Gtk2->main();

0;

lib/Gtk3/MIDIPlot.pm
# SeekMIDI, a simple graphical MIDI sequencer
# Copyright (C) 2016  oldtechaa  <oldtechaa@gmail.com>
# Version 0.1.0

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# ALWAYS MAINTAIN NO WARNINGS AND NO ERRORS!

# custom widget class; separate from main package below
package Gtk3::MIDIPlot;

use warnings;
use strict;

# invoke dependency modules
use Gtk3;
use Pango;
use Readonly;
use base 'Gtk3::Box';

our $VERSION = 0.03;

Readonly my $FALSE => 0;
Readonly my $TRUE  => 1;

Readonly my $NOTE_ENABLED   => 0;
Readonly my $NOTE_CONTINUED => 1;
Readonly my $NOTE_START     => 2;
Readonly my $NOTE_LENGTH    => 3;
Readonly my $NOTE_VOLUME    => 4;

# set up black-white key pattern on left sidebar
Readonly::Array my @KEYS => (0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
                             0, 1, 0, 1, 0, 0, 1, 0,);

# makes a package-global array that holds note objects, and the global drawing area
my @notes;

# set up package-global variables for widgets that need to be accessed throughout the package
my $scroll;
my $volume_slider;
my $volume_reveal;

# initialize the drag variables to a no-drag state
my ($drag_row, $drag_start, $drag_mode) = (-1, -1, -1);

# set up single note selection
my @select_single = (-1, -1);

# set some global constants
my ($cell_width, $cell_height, $number_of_cells, $cell_time, $default_volume) = (12, 8, 2400, 6, 127);

# sets up the class; asks for the signals we need; sets main widget size
sub new {
    my $canvas = Gtk3::DrawingArea->new();
    my $this = bless Gtk3::Box->new('horizontal', 6), shift;
    $scroll = Gtk3::ScrolledWindow->new();
    my $vbox = Gtk3::Box->new('vertical', 6);
    my $volume_label = Gtk3::Label->new('Vol: ');
    $volume_slider = Gtk3::Scale->new_with_range('vertical', 0, 127, 1);
    $volume_reveal = Gtk3::Revealer->new();

    $volume_slider->set_inverted(1);

    $this->pack_start($scroll, $TRUE, $TRUE, 0);
    $this->pack_start($volume_reveal, $FALSE, $FALSE, 0);

    $scroll->add_with_viewport($canvas);

    $vbox->pack_start($volume_label, $FALSE, $FALSE, 0);
    $vbox->pack_start($volume_slider, $TRUE, $TRUE, 0);
    $volume_reveal->add($vbox);
    $volume_reveal->set_transition_type('slide_right');

    $canvas->signal_connect('draw' => 'Gtk3::MIDIPlot::refresh');
    $canvas->signal_connect('button_press_event' => 'Gtk3::MIDIPlot::button');
    $canvas->signal_connect('motion_notify_event' => 'Gtk3::MIDIPlot::motion');
    $canvas->signal_connect('button_release_event' => 'Gtk3::MIDIPlot::release');
    $canvas->signal_connect('realize' => sub{$canvas->get_window()->set_events(['button-press-mask', 'button-motion-mask', 'button-release-mask'])});

    $scroll->get_hadjustment->signal_connect('value_changed' => sub {$canvas->queue_draw()});
    $scroll->get_vadjustment->signal_connect('value_changed' => sub {$canvas->queue_draw()});

    $volume_slider->get_adjustment->signal_connect('value_changed' => 'Gtk3::MIDIPlot::volume_changed');

    $canvas->set_size_request(($number_of_cells + 3) * $cell_width, 130 * $cell_height);

    return $this;
}

# refresh handler; handles drawing grid and objects
# NOTE: $xmin, $ymin refer to grid area coordinates, NOT global drawing area coordinates. 3 cells on the left and 2 on top are taken by sidebar and header
sub refresh {
    # gets Cairo context
    my ($this, $cairo) = @_;

    # sets drawing color for main grid
    $cairo->set_source_rgb(0.75, 0.75, 0.75);

    # get the current scroll positions and size of the window, then convert to grid-blocks, adjusting to draw surrounding blocks also, and make sure we don't go out of bounds
    my ($xmin, $ymin, $xmax, $ymax) = (int(($cairo->clip_extents())[0] / $cell_width) + 3, int(($cairo->clip_extents())[1] / $cell_height) + 2, int(($cairo->clip_extents())[2] / $cell_width), int(($cairo->clip_extents())[3] / $cell_height));

    # these two loops create the background grid
    for ($xmin .. $xmax) {
        $cairo->move_to($_ * $cell_width, $ymin * $cell_height);
        $cairo->line_to($_ * $cell_width, $ymax * $cell_height);
    }
    for ($ymin .. $ymax) {
        $cairo->move_to($xmin * $cell_width, $_ * $cell_height);
        $cairo->line_to($xmax * $cell_width, $_ * $cell_height);
    }

    # the grid must be drawn before we start redrawing the key objects
    $cairo->stroke();

    $cairo->set_source_rgb(0.5, 0.5, 0.5);

    # set up Pango for time header
    my $pango = Pango::Cairo::create_layout($cairo);
    my $pango_attributes = Pango::AttrList->new();
    $pango_attributes->insert(Pango::AttrSize->new(8192));
    $pango->set_attributes($pango_attributes);

    # draw time header and darker lines on time divisions
    for ($xmin - 3 .. $xmax - 3) {
        if ($_ % (96 / $cell_time) == 0) {
            $cairo->move_to(($_ + 3) * $cell_width, $ymin * $cell_height);
            $cairo->line_to(($_ + 3) * $cell_width, $ymax * $cell_height);

            $pango->set_text($_ / (96 / $cell_time));
            my ($pango_width, $pango_height) = $pango->get_size();
            $cairo->move_to($_ * $cell_width - $pango_width / Pango->scale() / 2 + 3 * $cell_width, $ymin * $cell_height - ($cell_height * 1.5));
            Pango::Cairo::show_layout($cairo, $pango);
        }
    }

    # draw darker top and bottom lines
    if ($ymin == 2) {
        $cairo->move_to($xmin * $cell_width, 2 * $cell_height);
        $cairo->line_to($xmax * $cell_width, 2 * $cell_height);
    }
    if ($ymax == 130) {
        $cairo->move_to($xmin * $cell_width, 130 * $cell_height);
        $cairo->line_to($xmax * $cell_width, 130 * $cell_height);
    }

    # stroke the darker lines
    $cairo->stroke();

    # create the piano key sidebar
    $cairo->set_source_rgb(0, 0, 0);
    for ($ymin .. $ymax - 1) {
        if ($KEYS[127 - ($_ - 2)] == $TRUE) {
            $cairo->rectangle(($xmin - 3) * $cell_width, $_ * $cell_height + 1, ($cell_width * 3) * 0.8, $cell_height * 0.75);
        }
    }

    # this checks for events with their state set to true, then draws them, scanning the leftmost column first, then all others
    for ($ymin .. $ymax - 1) {
        if (is_enabled($xmin - 3, $_ - 2)) {
            my $start_note = $notes[$xmin - 3][$_ - 2][$NOTE_START];
            $cairo->rectangle($xmin * $cell_width, $_ * $cell_height, ($notes[$start_note][$_ - 2][$NOTE_LENGTH] - (($xmin - 3) - $start_note)) * $cell_width, $cell_height);
        }
    }
    for my $incx ($xmin + 1 .. $xmax - 1) {
        for my $incy ($ymin .. $ymax - 1) {
            if (is_enabled($incx - 3, $incy - 2) && $notes[$incx - 3][$incy - 2][$NOTE_CONTINUED] == $FALSE) {
                $cairo->rectangle($incx * $cell_width, $incy * $cell_height, $notes[$incx - 3][$incy - 2][$NOTE_LENGTH] * $cell_width, $cell_height);
            }
        }
    }

    # fill applies the black source onto the destination through the mask with rectangular holes
    $cairo->fill();

    return;
}

# handles mouse-clicks on the custom widget
sub button {
    my ($this, $event) = @_;

    my ($xcell, $ycell) = (($event->x - ($event->x % $cell_width)) / $cell_width, ($event->y - ($event->y % $cell_height)) / $cell_height);
    my ($xmin, $ymin) = (int($scroll->get_hadjustment()->get_value() / $cell_width) + 3, int($scroll->get_vadjustment()->get_value() / $cell_height) + 2);
    my ($x, $y) = ($xcell - 3, $ycell - 2);

    my $max_cell = $number_of_cells - 1;

    # left mouse button
    if ($event->button == 1) {
        if ($xcell >= $xmin && $ycell >= $ymin) {
            if (is_enabled($x, $y)) {
                # select a note
                select_note($notes[$x][$y][$NOTE_START], $y);
            } else {
                # add a note and refresh
                add_note($x, $y);
                $this->queue_draw();

                $drag_mode = 0;
                $drag_row = $y;
            }
        }
    # right mouse button
    } elsif ($event->button == 3) {
        if ($xcell >= $xmin && $ycell >= $ymin) {
            if (is_enabled($x, $y)) {
                # remove the note
                if (@select_single = ($notes[$x][$y][$NOTE_START], $y)) {@select_single = (-1, -1)};

                for ($notes[$x][$y][$NOTE_START] .. $notes[$x][$y][$NOTE_START] + $notes[$notes[$x][$y][$NOTE_START]][$y][$NOTE_LENGTH] - 1) {
                    $notes[$_][$y][$NOTE_ENABLED] = $FALSE;
                }
            } else {
                # turn off selection
                @select_single = (-1, -1);
            }
            # hide volume slider
            $volume_reveal->set_reveal_child($FALSE);
            $volume_reveal->hide();
            $this->queue_draw();
        }
    }

    return;
}

# handles mouse drag across the widget
sub motion {
    my ($this, $event) = @_;

    my ($xcell, $ycell) = (($event->x - ($event->x % $cell_width)) / $cell_width, ($event->y - ($event->y % $cell_height)) / $cell_height);
    my ($xmin, $ymin) = (int($scroll->get_hadjustment()->get_value() / $cell_width) + 3, int($scroll->get_vadjustment()->get_value() / $cell_height) + 2);
    my ($x, $y) = ($xcell - 3, $ycell - 2);

    # check if the underlying cell is set or not and if not, check which mouse button is pressed, then set $notes and refresh
    if ($drag_mode == 0) {
        if ($xcell >= $xmin) {
            add_note($x, $drag_row);
            $this->queue_draw();
        }
    }

    return;
}

# clears the current drag row and mode when the drag is ended
sub release {
    ($drag_row, $drag_mode) = (-1, -1);

    return;
}

# creates a new note and selects it
sub add_note {
    my ($x, $y) = @_;

    @{$notes[$x][$y]}[$NOTE_ENABLED, $NOTE_CONTINUED, $NOTE_START] = ($TRUE, $FALSE, $x);
    if (is_enabled($x - 1, $y)) {
        @{$notes[$x][$y]}[$NOTE_CONTINUED, $NOTE_START] = ($TRUE, $notes[$x - 1][$y][$NOTE_START]);
    }
    for ($notes[$x][$y][$NOTE_START] + 1 .. $number_of_cells) {
        if (is_enabled($_, $y)) {
            @{$notes[$_][$y]}[$NOTE_CONTINUED, $NOTE_START] = ($TRUE, $notes[$x][$y][$NOTE_START]);
        } else {
            $notes[$notes[$x][$y][$NOTE_START]][$y][$NOTE_LENGTH] = $_  - $notes[$x][$y][$NOTE_START];
            last;
        }
    }

    select_note($notes[$x][$y][$NOTE_START], $y);

    return;
}

# sets a note's selected attribute to true
sub select_note {
	my ($x, $y) = @_;
	@select_single = ($x, $y);

    # show volume slider for selected note
    $volume_reveal->show();
    $volume_reveal->set_reveal_child(1);
    if (!$notes[$notes[$x][$y][$NOTE_START]][$y][$NOTE_VOLUME]) {
        $notes[$notes[$x][$y][$NOTE_START]][$y][$NOTE_VOLUME] = $default_volume;
    }
    $volume_slider->get_adjustment()->set_value($notes[$notes[$x][$y][$NOTE_START]][$y][$NOTE_VOLUME]);

    return;
}

# changes volume attribute of a note
sub volume_changed {
    $notes[$select_single[0]][$select_single[1]][$NOTE_VOLUME] = shift->get_value();

    return;
}

# gets the actual MIDI data for output
sub get_data {
    my ($patch) = @_;

    my @events;

    push @events, ['patch_change', 0, 0, $patch];

    for (0 .. 127) {
        if (is_enabled(0, $_)) {
            push @events, ['note_on', 0, 0, 127 - $_, $notes[0][$_][$NOTE_VOLUME]];
        }
    }
    my $delta = $cell_time;
    for my $incx (1 .. $number_of_cells) {
        for my $incy (0 .. 127) {
            if (!is_enabled($incx, $incy) && is_enabled($incx - 1, $incy)) {
                push @events, ['note_off', $delta, 0, 127 - $incy, $notes[$notes[$incx - 1][$incy][$NOTE_START]][$incy][$NOTE_VOLUME]];
                $delta = 0;
            } elsif (is_enabled($incx, $incy) && !is_enabled($incx - 1, $incy)) {
                push @events, ['note_on', $delta, 0, 127 - $incy, $notes[$incx][$incy][$NOTE_VOLUME]];
                $delta = 0;
            }
        }
        $delta += $cell_time;
    }

    return \@events;
}

# makes it so we can get the vbox from outside the package
sub get_volume_reveal {
    return $volume_reveal;
}

# tells us whether a cell contains (part of) a note
sub is_enabled {
    my ($x, $y) = @_;
    if ($notes[$x][$y][$NOTE_ENABLED]) {
        return $TRUE;
    } else {
        return $FALSE;
    }
}

1;
